File: packages/api-client/src/namo.ts
import type { ApiClient } from "./index";

export interface TempleRead {
  id: string;
  city_id: string;
  state_id: string;
  country_id: string;
  name: string;
  slug: string;
  temple_type: string;
  deity_name: string | null;
  address_line1: string;
  address_line2: string | null;
  pincode: string;
  latitude: number | null;
  longitude: number | null;
  description: string | null;
  dress_code: string | null;
  website_url: string | null;
  phone_number: string | null;
  is_active: boolean;
  rating_avg: number;
  review_count: number;
}

export interface TempleDetailRead extends TempleRead {
  images: Array<{ id: string; caption: string | null; is_primary: boolean }>;
  timings: Array<{ id: string; day_of_week: number; opens_at: string; closes_at: string; notes: string | null }>;
  festivals: FestivalRead[];
  events: Array<{ id: string; title: string; description: string | null; starts_on: string }>;
  attractions: Array<{ id: string; name: string; distance_km: number | null; description: string | null }>;
  reviews: TempleReviewRead[];
}

export interface TempleListResponse {
  items: TempleRead[];
  page: number;
  page_size: number;
  total: number;
  pages: number;
}

export interface TempleSearchResponse extends TempleListResponse {
  suggestions: string[];
}

export interface DarshanSlotRead {
  id: string;
  temple_id: string;
  slot_date: string;
  start_time: string;
  end_time: string;
  capacity: number;
  booked_count: number;
  available_count: number;
  slot_status: string;
}

export interface DarshanBookingRead {
  id: string;
  booking_number: string;
  temple_id: string;
  darshan_slot_id: string | null;
  booking_status: string;
  visit_date: string;
  party_size: number;
  notes: string | null;
  qr_ticket?: string;
}

export interface PujaRead {
  id: string;
  temple_id: string;
  title: string;
  description: string | null;
  base_price: number;
  status: string;
}

export interface DonationRead {
  id: string;
  temple_id: string;
  donor_name: string;
  purpose: string | null;
  amount: number;
  currency: string;
  receipt_number: string | null;
  created_at: string;
}

export interface AccommodationRead {
  id: string;
  temple_id: string | null;
  name: string;
  accommodation_type: string;
  address_line1: string;
  is_active: boolean;
}

export interface RoomRead {
  id: string;
  hotel_id: string;
  room_type: string;
  capacity: number;
  price_per_night: number;
  is_available: boolean;
}

export interface TravelPackageRead {
  id: string;
  temple_id: string | null;
  title: string;
  description: string | null;
  price: number;
  currency: string;
}

export interface TourGuideRead {
  id: string;
  name: string;
  rating_avg: number;
  phone_number: string | null;
}

export interface TransportationRead {
  id: string;
  provider_name: string;
  transport_type: string;
  temple_id: string | null;
}

export interface TripPlannerResponse {
  summary: string;
  days: Array<Record<string, unknown>>;
  estimated_budget: number;
  recommendations: string[];
}

export interface AIAssistantResponse {
  answer: string;
  suggested_actions: string[];
  sources: string[];
}

export interface FestivalRead {
  id: string;
  temple_id: string;
  name: string;
  description: string | null;
  starts_on: string;
  ends_on: string | null;
}

export interface FestivalReminderRead {
  festival_id: string;
  temple_id: string;
  title: string;
  starts_on: string;
  reminder_text: string;
}

export interface TempleReviewRead {
  id: string;
  rating: number;
  comment: string | null;
  user_id: string;
}

export interface AnalyticsSummary {
  temples: number;
  darshan_bookings: number;
  puja_bookings: number;
  donations: number;
  donation_amount: number;
  accommodation_bookings: number;
  travel_bookings: number;
  reviews: number;
}

export interface UserRead {
  id: string;
  email: string;
  full_name: string | null;
  is_active: boolean;
  is_verified: boolean;
}

export interface AuthLoginResponse {
  access_token: string;
  token_type: string;
}

export interface AuthRegisterPayload {
  email: string;
  password: string;
  full_name: string;
}

export interface AuthForgotPasswordResponse {
  message: string;
}

export interface AuthResetPasswordPayload {
  token: string;
  new_password: string;
}

export interface AuthVerifyEmailResponse {
  message: string;
}

export interface SessionRead {
  id: string;
  device_info: string | null;
  ip_address: string | null;
  last_active: string;
  is_current: boolean;
}

export interface NotificationRead {
  id: string;
  title: string;
  body: string;
  channel: string;
  is_read: boolean;
  created_at: string;
}

export interface NotificationListResponse {
  items: NotificationRead[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
  unread_count: number;
}

export interface TwoFactorSetupResponse {
  secret: string;
  qr_code_url: string;
  recovery_codes: string[];
}

export interface ChangePasswordPayload {
  current_password: string;
  new_password: string;
}

export function createNamoApi(client: ApiClient) {
  const base = "/namo";

  return {
    listTemples(params?: { page?: number; page_size?: number; search?: string; category?: string }) {
      const query = new URLSearchParams();
      if (params?.page) query.set("page", String(params.page));
      if (params?.page_size) query.set("page_size", String(params.page_size));
      if (params?.search) query.set("search", params.search);
      if (params?.category) query.set("category", params.category);
      const qs = query.toString();
      return client.request<TempleListResponse>(`${base}/temples${qs ? `?${qs}` : ""}`);
    },

    getTemple(templeId: string) {
      return client.request<TempleDetailRead>(`${base}/temples/${templeId}`);
    },

    createTemple(payload: Record<string, unknown>) {
      return client.request<TempleRead>(`${base}/temples`, { method: "POST", body: JSON.stringify(payload) });
    },

    updateTemple(templeId: string, payload: Record<string, unknown>) {
      return client.request<TempleRead>(`${base}/temples/${templeId}`, { method: "PATCH", body: JSON.stringify(payload) });
    },

    deleteTemple(templeId: string) {
      return client.request<{ message: string; data: string }>(`${base}/temples/${templeId}`, { method: "DELETE" });
    },

    searchTemples(params?: { q?: string; category?: string; page?: number; page_size?: number }) {
      const query = new URLSearchParams();
      if (params?.q) query.set("q", params.q);
      if (params?.category) query.set("category", params.category);
      if (params?.page) query.set("page", String(params.page));
      if (params?.page_size) query.set("page_size", String(params.page_size));
      const qs = query.toString();
      return client.request<TempleSearchResponse>(`${base}/search/temples${qs ? `?${qs}` : ""}`);
    },

    nearbyTemples(payload: { latitude: number; longitude: number; radius_km?: number }) {
      return client.request<TempleRead[]>(`${base}/search/nearby`, { method: "POST", body: JSON.stringify(payload) });
    },

    popularTemples(limit = 10) {
      return client.request<TempleRead[]>(`${base}/search/popular?limit=${limit}`);
    },

    listDarshanSlots(templeId: string) {
      return client.request<DarshanSlotRead[]>(`${base}/temples/${templeId}/darshan/slots`);
    },

    bookDarshan(payload: { temple_id: string; darshan_slot_id?: string; visit_date: string; party_size: number; notes?: string }) {
      return client.request<DarshanBookingRead>(`${base}/darshan/bookings`, { method: "POST", body: JSON.stringify(payload) });
    },

    myDarshanBookings() {
      return client.request<DarshanBookingRead[]>(`${base}/darshan/bookings`);
    },

    cancelDarshanBooking(bookingId: string, reason?: string) {
      return client.request<DarshanBookingRead>(`${base}/darshan/bookings/${bookingId}/cancel`, {
        method: "POST",
        body: JSON.stringify({ reason }),
      });
    },

    listPujaPackages(templeId: string) {
      return client.request<PujaRead[]>(`${base}/temples/${templeId}/puja/packages`);
    },

    createPuja(payload: Record<string, unknown>) {
      return client.request<PujaRead>(`${base}/puja/packages`, { method: "POST", body: JSON.stringify(payload) });
    },

    bookPuja(payload: Record<string, unknown>) {
      return client.request(`${base}/puja/bookings`, { method: "POST", body: JSON.stringify(payload) });
    },

    myPujaBookings() {
      return client.request(`${base}/puja/bookings`);
    },

    donate(payload: { temple_id: string; donor_name: string; purpose?: string; amount: number; currency?: string; provider?: string }) {
      return client.request<DonationRead>(`${base}/donations`, { method: "POST", body: JSON.stringify(payload) });
    },

    myDonations() {
      return client.request<DonationRead[]>(`${base}/donations`);
    },

    listAccommodation(templeId?: string) {
      const qs = templeId ? `?temple_id=${templeId}` : "";
      return client.request<AccommodationRead[]>(`${base}/accommodation${qs}`);
    },

    listRooms(hotelId?: string) {
      const qs = hotelId ? `?hotel_id=${hotelId}` : "";
      return client.request<RoomRead[]>(`${base}/accommodation/rooms${qs}`);
    },

    bookAccommodation(payload: Record<string, unknown>) {
      return client.request(`${base}/accommodation/bookings`, { method: "POST", body: JSON.stringify(payload) });
    },

    listTravelPackages(templeId?: string) {
      const qs = templeId ? `?temple_id=${templeId}` : "";
      return client.request<TravelPackageRead[]>(`${base}/travel/packages${qs}`);
    },

    listGuides(cityId?: string) {
      const qs = cityId ? `?city_id=${cityId}` : "";
      return client.request<TourGuideRead[]>(`${base}/travel/guides${qs}`);
    },

    listTransport(templeId?: string) {
      const qs = templeId ? `?temple_id=${templeId}` : "";
      return client.request<TransportationRead[]>(`${base}/travel/transport${qs}`);
    },

    planTrip(payload: Record<string, unknown>) {
      return client.request<TripPlannerResponse>(`${base}/travel/planner`, { method: "POST", body: JSON.stringify(payload) });
    },

    bookTravel(payload: Record<string, unknown>) {
      return client.request(`${base}/travel/bookings`, { method: "POST", body: JSON.stringify(payload) });
    },

    aiRecommendations(message: string, templeId?: string) {
      return client.request<AIAssistantResponse>(`${base}/search/recommendations`, {
        method: "POST",
        body: JSON.stringify({ message, temple_id: templeId }),
      });
    },

    spiritualGuide(message: string, templeId?: string) {
      return client.request<AIAssistantResponse>(`${base}/ai/spiritual-guide`, {
        method: "POST",
        body: JSON.stringify({ message, temple_id: templeId }),
      });
    },

    voiceAssistant(message: string, transcript?: string, templeId?: string) {
      return client.request<AIAssistantResponse>(`${base}/ai/voice-assistant`, {
        method: "POST",
        body: JSON.stringify({ message, transcript, temple_id: templeId }),
      });
    },

    faqAssistant(message: string, templeId?: string) {
      return client.request<AIAssistantResponse>(`${base}/ai/faq`, {
        method: "POST",
        body: JSON.stringify({ message, temple_id: templeId }),
      });
    },

    nearbyAi(message: string, templeId?: string) {
      return client.request<AIAssistantResponse>(`${base}/ai/nearby`, {
        method: "POST",
        body: JSON.stringify({ message, temple_id: templeId }),
      });
    },

    festivalReminders(payload?: { temple_id?: string; state_id?: string; days_ahead?: number }) {
      return client.request<FestivalReminderRead[]>(`${base}/ai/festival-reminders`, {
        method: "POST",
        body: JSON.stringify(payload ?? {}),
      });
    },

    createFestival(templeId: string, payload: Record<string, unknown>) {
      return client.request<FestivalRead>(`${base}/temples/${templeId}/festivals`, {
        method: "POST",
        body: JSON.stringify(payload),
      });
    },

    analyticsSummary() {
      return client.request<AnalyticsSummary>(`${base}/analytics/summary`);
    },

    adminUsers() {
      return client.request<UserRead[]>("/admin/users");
    },

    // Auth
    login(email: string, password: string) {
      return client.request<AuthLoginResponse>("/auth/login", {
        method: "POST",
        body: JSON.stringify({ email, password }),
      });
    },

    register(payload: AuthRegisterPayload) {
      return client.request<AuthLoginResponse>("/auth/register", {
        method: "POST",
        body: JSON.stringify(payload),
      });
    },

    forgotPassword(email: string) {
      return client.request<AuthForgotPasswordResponse>("/auth/forgot-password", {
        method: "POST",
        body: JSON.stringify({ email }),
      });
    },

    resetPassword(payload: AuthResetPasswordPayload) {
      return client.request<{ message: string }>("/auth/reset-password", {
        method: "POST",
        body: JSON.stringify(payload),
      });
    },

    verifyEmail(token: string) {
      return client.request<AuthVerifyEmailResponse>(`/auth/verify-email?token=${encodeURIComponent(token)}`, {
        method: "POST",
      });
    },

    // Sessions
    listSessions() {
      return client.request<SessionRead[]>("/auth/sessions");
    },

    revokeSession(sessionId: string) {
      return client.request<{ message: string }>(`/auth/sessions/${sessionId}`, {
        method: "DELETE",
      });
    },

    revokeAllOtherSessions() {
      return client.request<{ message: string }>("/auth/sessions", {
        method: "DELETE",
      });
    },

    // Notifications
    listNotifications(params?: { page?: number; page_size?: number; channel?: string; unread_only?: boolean }) {
      const query = new URLSearchParams();
      if (params?.page) query.set("page", String(params.page));
      if (params?.page_size) query.set("page_size", String(params.page_size));
      if (params?.channel) query.set("channel", params.channel);
      if (params?.unread_only) query.set("unread_only", String(params.unread_only));
      const qs = query.toString();
      return client.request<NotificationListResponse>(`/notifications${qs ? `?${qs}` : ""}`);
    },

    markNotificationRead(notificationId: string) {
      return client.request<{ message: string }>(`/notifications/${notificationId}/read`, {
        method: "POST",
      });
    },

    markAllNotificationsRead() {
      return client.request<{ message: string }>("/notifications/read-all", {
        method: "POST",
      });
    },

    // Security
    changePassword(payload: ChangePasswordPayload) {
      return client.request<{ message: string }>("/auth/change-password", {
        method: "POST",
        body: JSON.stringify(payload),
      });
    },

    enableTwoFactor() {
      return client.request<TwoFactorSetupResponse>("/auth/2fa/enable", {
        method: "POST",
      });
    },

    disableTwoFactor(code: string) {
      return client.request<{ message: string }>("/auth/2fa/disable", {
        method: "POST",
        body: JSON.stringify({ code }),
      });
    },
  };
}

export type NamoApi = ReturnType<typeof createNamoApi>;

File: backend/app/schemas/namo.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field

from backend.app.schemas.common import ORMModel


class PageParams(BaseModel):
    """Shared pagination and search controls for Namo Setu list endpoints."""

    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    search: str | None = None


class NamoPage(BaseModel):
    page: int
    page_size: int
    total: int
    pages: int


class TempleBase(BaseModel):
    city_id: str
    state_id: str
    country_id: str
    address_id: str | None = None
    name: str
    slug: str
    temple_type: str = "main"
    deity_name: str | None = None
    address_line1: str
    address_line2: str | None = None
    pincode: str
    latitude: float | None = None
    longitude: float | None = None
    description: str | None = None
    history: str | None = None
    dress_code: str | None = None
    accessibility_notes: str | None = None
    website_url: str | None = None
    phone_number: str | None = None
    is_active: bool = True


class TempleCreate(TempleBase):
    """Admin request for creating a temple profile."""


class TempleUpdate(BaseModel):
    name: str | None = None
    slug: str | None = None
    temple_type: str | None = None
    deity_name: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    pincode: str | None = None
    latitude: float | None = None
    longitude: float | None = None
    description: str | None = None
    history: str | None = None
    dress_code: str | None = None
    accessibility_notes: str | None = None
    website_url: str | None = None
    phone_number: str | None = None
    is_active: bool | None = None


class TempleRead(ORMModel):
    id: str
    city_id: str
    state_id: str
    country_id: str
    address_id: str | None = None
    name: str
    slug: str
    temple_type: str
    deity_name: str | None = None
    address_line1: str
    address_line2: str | None = None
    pincode: str
    latitude: float | None = None
    longitude: float | None = None
    description: str | None = None
    history: str | None = None
    dress_code: str | None = None
    accessibility_notes: str | None = None
    website_url: str | None = None
    phone_number: str | None = None
    is_active: bool
    created_at: datetime
    rating_avg: float = 0
    review_count: int = 0


class TempleListResponse(NamoPage):
    items: list[TempleRead]


class TempleImageCreate(BaseModel):
    media_id: str
    caption: str | None = None
    sort_order: int = 0
    is_primary: bool = False


class TempleImageRead(ORMModel):
    id: str
    temple_id: str
    media_id: str
    caption: str | None = None
    sort_order: int
    is_primary: bool


class TempleTimingCreate(BaseModel):
    day_of_week: int = Field(ge=0, le=6)
    opens_at: str
    closes_at: str
    is_closed: bool = False
    notes: str | None = None


class TempleTimingRead(ORMModel):
    id: str
    temple_id: str
    day_of_week: int
    opens_at: str
    closes_at: str
    is_closed: bool
    notes: str | None = None


class TempleEventCreate(BaseModel):
    title: str
    description: str | None = None
    starts_on: datetime
    ends_on: datetime | None = None
    is_public: bool = True


class TempleEventRead(ORMModel):
    id: str
    temple_id: str
    title: str
    description: str | None = None
    starts_on: datetime
    ends_on: datetime | None = None
    is_public: bool


class FestivalCreate(BaseModel):
    name: str
    description: str | None = None
    starts_on: datetime
    ends_on: datetime | None = None
    annual_recurring: bool = True


class FestivalRead(ORMModel):
    id: str
    temple_id: str
    name: str
    description: str | None = None
    starts_on: datetime
    ends_on: datetime | None = None
    annual_recurring: bool


class NearbyAttractionCreate(BaseModel):
    name: str
    category: str = "landmark"
    description: str | None = None
    distance_km: float | None = None
    duration_minutes: int | None = None
    latitude: float | None = None
    longitude: float | None = None
    is_active: bool = True


class NearbyAttractionRead(ORMModel):
    id: str
    temple_id: str
    name: str
    category: str
    description: str | None = None
    distance_km: float | None = None
    duration_minutes: int | None = None
    latitude: float | None = None
    longitude: float | None = None
    is_active: bool


class TempleReviewCreate(BaseModel):
    rating: int = Field(ge=1, le=5)
    title: str | None = None
    body: str | None = None


class TempleReviewRead(ORMModel):
    id: str
    temple_id: str
    user_id: str | None = None
    rating: int
    title: str | None = None
    body: str | None = None
    is_verified: bool
    created_at: datetime


class TempleDetailRead(TempleRead):
    images: list[TempleImageRead] = Field(default_factory=list)
    timings: list[TempleTimingRead] = Field(default_factory=list)
    festivals: list[FestivalRead] = Field(default_factory=list)
    events: list[TempleEventRead] = Field(default_factory=list)
    attractions: list[NearbyAttractionRead] = Field(default_factory=list)
    reviews: list[TempleReviewRead] = Field(default_factory=list)


class TempleSearchResponse(NamoPage):
    items: list[TempleRead]
    suggestions: list[str] = Field(default_factory=list)


class NearbyTempleRequest(BaseModel):
    latitude: float
    longitude: float
    radius_km: float = Field(default=50, gt=0, le=500)


class DarshanSlotCreate(BaseModel):
    temple_id: str
    slot_date: date
    start_time: str
    end_time: str
    capacity: int = Field(default=0, ge=0)
    slot_status: str = "available"


class DarshanSlotUpdate(BaseModel):
    slot_date: date | None = None
    start_time: str | None = None
    end_time: str | None = None
    capacity: int | None = Field(default=None, ge=0)
    slot_status: str | None = None


class DarshanSlotRead(ORMModel):
    id: str
    temple_id: str
    slot_date: date
    start_time: str
    end_time: str
    capacity: int
    booked_count: int
    slot_status: str
    available_count: int = 0


class DarshanBookingCreate(BaseModel):
    temple_id: str
    darshan_slot_id: str | None = None
    visit_date: date
    party_size: int = Field(default=1, ge=1, le=25)
    notes: str | None = None


class DarshanBookingRead(ORMModel):
    id: str
    booking_number: str
    user_id: str
    temple_id: str
    darshan_slot_id: str | None = None
    booking_status: str
    visit_date: date
    party_size: int
    notes: str | None = None
    total_amount: float
    currency: str
    qr_ticket: str = ""


class BookingCancellationRequest(BaseModel):
    reason: str | None = None


class PujaCreate(BaseModel):
    temple_id: str
    title: str
    description: str | None = None
    status: str = "scheduled"
    base_price: float = Field(default=0, ge=0)


class PujaRead(ORMModel):
    id: str
    temple_id: str
    title: str
    description: str | None = None
    status: str
    base_price: float


class PanditCreate(BaseModel):
    temple_id: str
    name: str
    phone_number: str | None = None
    language_id: str | None = None
    is_active: bool = True


class PanditRead(ORMModel):
    id: str
    temple_id: str
    name: str
    phone_number: str | None = None
    language_id: str | None = None
    is_active: bool


class PujaBookingCreate(BaseModel):
    puja_id: str
    temple_id: str
    pandit_id: str | None = None
    booking_date: date
    scheduled_at: datetime | None = None
    devotee_name: str | None = None
    notes: str | None = None


class PujaBookingRead(ORMModel):
    id: str
    puja_id: str
    user_id: str
    temple_id: str
    pandit_id: str | None = None
    booking_date: date
    scheduled_at: datetime | None = None
    status: str
    amount: float
    notes: str | None = None
    devotee_name: str | None = None


class DonationCreate(BaseModel):
    temple_id: str
    donor_name: str
    purpose: str | None = None
    amount: float = Field(gt=0)
    currency: str = "INR"
    provider: str = "online"


class DonationRead(ORMModel):
    id: str
    temple_id: str
    user_id: str | None = None
    donation_reference: str
    donor_name: str
    purpose: str | None = None
    amount: float
    currency: str
    receipt_number: str = ""
    created_at: datetime


class AccommodationCreate(BaseModel):
    temple_id: str | None = None
    name: str
    accommodation_type: str = "hotel"
    address_id: str | None = None
    is_active: bool = True


class AccommodationRead(ORMModel):
    id: str
    temple_id: str | None = None
    name: str
    accommodation_type: str
    address_id: str | None = None
    is_active: bool


class HotelCreate(BaseModel):
    accommodation_id: str
    star_rating: int | None = Field(default=None, ge=0, le=5)
    check_in_time: str | None = None
    check_out_time: str | None = None
    contact_number: str | None = None
    is_active: bool = True


class HotelRead(ORMModel):
    id: str
    accommodation_id: str
    star_rating: int | None = None
    check_in_time: str | None = None
    check_out_time: str | None = None
    contact_number: str | None = None
    is_active: bool


class RoomCreate(BaseModel):
    hotel_id: str
    room_type: str
    capacity: int = Field(ge=1)
    price_per_night: float = Field(ge=0)
    is_available: bool = True
    is_active: bool = True


class RoomRead(ORMModel):
    id: str
    hotel_id: str
    room_type: str
    capacity: int
    price_per_night: float
    is_available: bool
    is_active: bool


class AccommodationBookingCreate(BaseModel):
    room_id: str
    temple_id: str | None = None
    check_in_date: date
    check_out_date: date
    guests: int = Field(default=1, ge=1)
    special_requests: str | None = None


class AccommodationBookingRead(ORMModel):
    id: str
    booking_reference: str
    user_id: str
    room_id: str
    temple_id: str | None = None
    check_in_date: date
    check_out_date: date
    guests: int
    status: str
    total_amount: float
    currency: str
    special_requests: str | None = None


class TransportationCreate(BaseModel):
    temple_id: str | None = None
    provider_name: str
    transport_type: str = "cab"
    vehicle_number: str | None = None
    contact_number: str | None = None
    is_active: bool = True


class TransportationRead(ORMModel):
    id: str
    temple_id: str | None = None
    provider_name: str
    transport_type: str
    vehicle_number: str | None = None
    contact_number: str | None = None
    is_active: bool


class PilgrimageRouteCreate(BaseModel):
    origin_city_id: str
    destination_city_id: str
    name: str
    description: str | None = None
    distance_km: float | None = None
    estimated_duration_minutes: int | None = None
    is_active: bool = True


class PilgrimageRouteRead(ORMModel):
    id: str
    origin_city_id: str
    destination_city_id: str
    name: str
    description: str | None = None
    distance_km: float | None = None
    estimated_duration_minutes: int | None = None
    is_active: bool


class TravelPackageCreate(BaseModel):
    temple_id: str | None = None
    title: str
    description: str | None = None
    price: float = Field(ge=0)
    currency: str = "INR"
    is_active: bool = True


class TravelPackageRead(ORMModel):
    id: str
    temple_id: str | None = None
    title: str
    description: str | None = None
    price: float
    currency: str
    is_active: bool


class TourGuideCreate(BaseModel):
    city_id: str | None = None
    name: str
    language_id: str | None = None
    phone_number: str | None = None
    rating_avg: float = Field(default=0, ge=0, le=5)
    is_active: bool = True


class TourGuideRead(ORMModel):
    id: str
    city_id: str | None = None
    name: str
    language_id: str | None = None
    phone_number: str | None = None
    rating_avg: float
    is_active: bool


class TravelPackageBookingCreate(BaseModel):
    package_id: str
    guide_id: str | None = None
    transport_id: str | None = None
    start_date: date
    travelers: int = Field(default=1, ge=1)
    notes: str | None = None


class TravelPackageBookingRead(ORMModel):
    id: str
    booking_reference: str
    user_id: str
    package_id: str
    guide_id: str | None = None
    transport_id: str | None = None
    start_date: date
    travelers: int
    status: str
    total_amount: float
    currency: str
    itinerary_json: str | None = None
    notes: str | None = None


class TripPlannerRequest(BaseModel):
    city_id: str | None = None
    temple_ids: list[str] = Field(default_factory=list)
    start_date: date
    days: int = Field(default=2, ge=1, le=14)
    travelers: int = Field(default=1, ge=1)
    interests: list[str] = Field(default_factory=list)
    pace: str = "balanced"


class TripPlannerResponse(BaseModel):
    summary: str
    days: list[dict[str, Any]]
    estimated_budget: float
    recommendations: list[str]


class AIAssistantRequest(BaseModel):
    message: str
    temple_id: str | None = None
    language: str = "en"
    user_context: dict[str, Any] = Field(default_factory=dict)


class AIAssistantResponse(BaseModel):
    answer: str
    suggested_actions: list[str] = Field(default_factory=list)
    sources: list[str] = Field(default_factory=list)


class VoiceAssistantRequest(AIAssistantRequest):
    transcript: str | None = None
    audio_url: str | None = None


class FestivalReminderRequest(BaseModel):
    temple_id: str | None = None
    state_id: str | None = None
    days_ahead: int = Field(default=30, ge=1, le=365)


class FestivalReminderRead(BaseModel):
    festival_id: str
    temple_id: str
    title: str
    starts_on: datetime
    reminder_text: str


class AnalyticsEventCreate(BaseModel):
    event_name: str
    event_type: str = "page_view"
    entity_type: str | None = None
    entity_id: str | None = None
    payload_json: str | None = None
    session_id: str | None = None


class AnalyticsSummary(BaseModel):
    temples: int
    darshan_bookings: int
    puja_bookings: int
    donations: int
    donation_amount: float
    accommodation_bookings: int
    travel_bookings: int
    reviews: int
